### PR TITLE
Support for include/exclude and multiple item patterns

### PR DESCRIPTION
**Pull request by gykuma**
Added support for multiple item patterns and include/exclude pattern for artifact download
Added L0 tests

### CODE REVIEW

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 1]
```diff
@@ -114,8 +115,33 @@ export class ArtifactEngine {
...
+                if (minimatch(path, (pattern.charAt(0) === '+') ? pattern.substr(1) : pattern, { dot: true, nocase: true })) {
+                    shouldDownload = true;
+                }
+            }
+        }
+        return shouldDownload;
+    }
+
+    createPatternList(artifactEngineOptions: ArtifactEngineOptions) {
+        if (artifactEngineOptions.itemPattern == null) {
```
Comment (omeshp) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: use !artifactEngineOptions.itemPattern

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 2]
```diff
@@ -114,8 +115,33 @@ export class ArtifactEngine {
         }
     }
 
+    shouldDownload(path: string) {
```
Comment (omeshp) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: Add a test to Integration tests also (https://github.com/Microsoft/vsts-rm-extensions/blob/master/Extensions/ArtifactEngine/IntegrationTests/jenkinsTests.ts)

Comment (gykuma) [reply Extensions/ArtifactEngine/Engine/artifactEngine.ts]: added tests

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ import { TicketState } from '../Models/ticketState';
 
 export class ArtifactEngine {
     processItems(sourceProvider: models.IArtifactProvider, destProvider: models.IArtifactProvider, artifactEngineOptions?: ArtifactEngineOptions): Promise<models.ArtifactDownloadTicket[]> {
+        this.createPatternList(artifactEngineOptions);
```
Comment (gykuma) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: move below initialization of options.

File: Extensions/ArtifactEngine/TestData/Jenkins/folder2/file1.pdb [hunk 4]
```diff
@@ -0,0 +1 @@
+dummyFile1Content
```
Comment (gykuma) [inline Extensions/ArtifactEngine/TestData/Jenkins/folder2/file1.pdb]: changes to testdata are only required for E2E tests.

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts (hunk 2 above)
Comment (omeshp) [reply Extensions/ArtifactEngine/Engine/artifactEngine.ts]: ok

**Review by omeshp — APPROVED** [review]

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts (hunk 2 above)
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: add return value to the signature?

shouldDownload(path: string): boolean

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 5]
```diff
@@ -114,8 +115,33 @@ export class ArtifactEngine {
         }
     }
 
+    shouldDownload(path: string) {
+        var shouldDownload = false;
+        for (let pattern of this.patternList) {
+            if (pattern.charAt(0) === '-' && (minimatch(path, pattern.substr(1), { dot: true, nocase: true }))) {
+                return false;
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: looks like exclude will override include in any case. this isn't true.

let's say, we have 2 specs:
- exclude c:\* 
- include c:\foo

i suppose we need to download c:\foo\bar, right?

can we do the following:

check if item matches any include
  - false if no include matches
if any include matches, check if item matches any exclude
 - true if no exclude matches
if any exclude matched check if include is a child of exclude 
 - true if include is child of exclude, else false

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 6]
```diff
@@ -114,8 +115,33 @@ export class ArtifactEngine {
...
+            else {
+                if (minimatch(path, (pattern.charAt(0) === '+') ? pattern.substr(1) : pattern, { dot: true, nocase: true })) {
+                    shouldDownload = true;
+                }
+            }
+        }
+        return shouldDownload;
+    }
+
+    createPatternList(artifactEngineOptions: ArtifactEngineOptions) {
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: good to add comment on what sort of itemPatterns are supported with examples
i suppose we can give the following examples:

a. none (in which case we assume **)
b. **
c. drop\starstar
d. "+drop\starstar","-drop\star.exe"
e. "+drop\foo\starstar","-drop\foo\bar\starstar"

PS: replace star with * (i had trouble adding ** as it was not showing up properly in comment)

File: Extensions/ArtifactEngine/Engine/artifactEngine.ts [hunk 7]
```diff
@@ -114,8 +115,33 @@ export class ArtifactEngine {
...
+        if (!artifactEngineOptions.itemPattern) {
+            this.patternList = ['**'];
+        }
+        else {
+            this.patternList = artifactEngineOptions.itemPattern.split('","').map((pattern) => pattern.replace(/"/g, ''));
```
Comment (GitHubSriramB) [inline Extensions/ArtifactEngine/Engine/artifactEngine.ts]: can pattern have space after comma? like:
"+drop\starstar", "-drop\start.exe" (notice space after comma)

Comment (gykuma) [reply Extensions/ArtifactEngine/Engine/artifactEngine.ts]: We planned the pattern string to be only as "+drop\**","-drop\folder\**" (without spaces). Should I add support for spaces also?

**Pull request merged**